Insight.Database/Reliable/RetryStrategy.cs

```diff
@@ -95,21 +95,38 @@ public TResult ExecuteWithRetry<TResult>(IDbCommand commandContext, Func<TResult
 			{
 				try
 				{
-					return func();
+                    return func();
 				}
 				catch (Exception ex)
 				{
 					// if it's not a transient error, then let it go
-					if (!IsTransientException(ex))
-						throw;
+				    if (!IsTransientException(ex))
+				    {
+				        commandContext.AutoClose();
+				        throw;
+				    }
 
 					// if the number of retries has been exceeded then throw
-					if (attempt >= MaxRetryCount)
-						throw;
+				    if (attempt >= MaxRetryCount)
+				    {
+                        commandContext.AutoClose();
+				        throw;
+				    }
 
 					// first off the event to tell someone that we are going to retry this
-					if (OnRetrying(commandContext, ex, attempt))
-						throw;
+				    if (OnRetrying(commandContext, ex, attempt))
+				    {
+                        commandContext.AutoClose();
+				        throw;
+				    }
+
+                    // if something went wrong and we currently have an open connection
+                    // release the open connection back to the connection pool
+                    // but only if not within a transaction, not sure what should be done?
+                    if (commandContext != null && commandContext.Transaction == null)

```

Comment:
Good catch. Transactions shouldn't be held open on transient errors, and the only way to do that would be to hold the connection open for longer. The best course of action is to abort the transaction and allow the system to retry the transaction using some other mechanism.

Maybe for best optionality, IRetryStrategy should support RetryInOptionTransactions and set the default value to false?

I'm ok with throwing when transactions are open, but we should add a note to the wiki.